proposals/pending-state.md

```diff
@@ -0,0 +1,105 @@
+
+# Goals
+
+* Allow components to communicate that they have pending asynchonous work to ancestors in the DOM tree.
+* Allow components to know if there is pending asynchronous work in descendants in the DOM tree.
+* Allow components to intercept, modify, and block pending state notifications.
+* Give guidance on the type of asynchronous work that should be notified.
+* Allow for some labelling or differentiation between types of asynchronous work.
+
+# Details
+
+## Asynchronous States
+
+There are four main states that an asynchronous operation can be in:
+
+* Not-started
+* Started
+* Sucessfully completed
+* Failed
+
+## The `pending-state` event
+
+Components with pending work indicate so by firing a composed, bubbling, `pending-state` Event, with a `promise` property:
+
+TypeScript interface:
+```ts
+interface PendingStateEvent extends Event {
+  promise: Promise<void>;
+}
+```
+
+Example:
+
+```js
+class PendingStateEvent extends Event {
+  constructor(promise) {
+    super('pending-state', {bubbles: true, composed: true});
+    this.promise = promise;
+  }
+}
+
+// Inside a component definition:
+class DoWorkElement extends HTMLElement {
+  async doWork() { /* ... */ }
+
+  startWork() {
+    this.dispatchEvent(new PendingStateEvent(this.doWork()));
+  }
+}
+
+// Inside a container component:
+class IndicateWorkElement extends HTMLElement {
+  constructor() {
+    super();
+    this.addEventListener('pending-state', async (e) => {
+      this.showSpinner();
+      await e.promise;
+      this.hideSpinner();
+    });
+  }
+}
+```
+
+The Promise must be resolved when the work is complete and rejected if the work fails. The value the Promise resolves to is unspecified.
+
+Using an event with a Promise allows us to represent three of the four asynchronous states:
+
+* Not-started: not represented
+* Started: `pending-state` event fired
+* Sucessfully completed: Promise resolved
+* Failed: Promise rejected
+
+# Open Questions
+
+## Should the event carry a Promise or should there be two events?
+
+Animations have multiple events: `animationstart` and `animationend`. Promises make it very easy to correlate the start of an operation with the end of that operation.
+
+## Event Name
+
+Is `pending-state` good? Should it be `pending-work`?
+
+## Types of Async Work
+
+There are some types of async work that we may not want to display UI affordances (like progress indicators) for. Async rendering used in order to yield to the browser's task queue for input hanlding and layout/paint for instance, should probably not trigger a progress indicator. Yet, code that measures style or layout may need to wait for rendering to complete.
+
+We could add a field to the event that indicates the type of work and standardize a small number of types, such as `loading` and `rendering`. An event to indicate async rendering starting and stopping has existing analogies with the `animationstart` and `animationend` events.
```

Comment:
passing on the type seems like a good idea 👍

would enable use cases like this measuring render timings and showing a spinner at the same time 👌
```js
this.addEventListener('pending-work', async (ev) => {
  if (ev.workType === 'loading') {
    this.showSpinner();
    await ev.promise;
    this.hideSpinner();
  } else if (ev.workType === 'rendering') {
    this.startRenderingMeasure();
    await ev.promise;
    this.stopRenderingMeasure();
  }
}
```

Reply:
I _suspect_ a type field is generally a good idea, but I'd want to get a lot of feedback from app builders on what types they'd actually use.

Reply:
@justinfagnani we currently use a variation of your implementation of pending-state container (mixin) in the lit-mail app. for now, we haven't had the need for types in the event handler, but I see the use case @daKmoR described.

Reply:
I'll open an issue for this.